kernel/src/sched/priority_scheduler.rs

```diff
@@ -117,6 +126,39 @@ impl<T: Sync + Send + PreemptSchedInfo + FromTask<U>, U: Sync + Send + CommonSch
     }
 }
 
+impl<T: Sync + Send + PreemptSchedInfo + FromTask<U>, U: Sync + Send + CommonSchedInfo>
+    SchedulerStats for PreemptScheduler<T, U>
+{
+    fn nr_queued(&self) -> u64 {
+        let _preempt_guard = disable_preempt();
+        let mut queued = 0;
+
+        for rq in self.rq.iter() {
+            let rq = rq.lock();
+
+            queued +=
+                rq.real_time_entities.len() + rq.normal_entities.len() + rq.lowest_entities.len();
+        }
+
+        queued as u64
+    }
+
+    fn nr_running(&self) -> u64 {
+        let _preempt_guard = disable_preempt();
+        let mut nr_running = 0;
+
+        for rq in self.rq.iter() {
+            let rq = rq.lock();
+
+            if rq.current.is_some() {
+                nr_running += 1;
+            }
+        }
+
+        nr_running
+    }
+}
```

Comment:
When the load average is queried through the proc file system, the two functions will be called one by one. And they will respectively lock the queues twice and unlock them twice. So there may be discrepancies if the current task is preempted in between the two calls. I think a good design of the trait won't introduce such problem. A suggestion is to return these two stats in one function.

It is a minor suggestion and feel free to argue with me.

Reply:
I believe you have a good point here. I'm actually thinking about what other things would need to get those statistics and if we would benefit from keeping them separated (maybe to save some precious cycles in a performance sensitive context ?)

It is to be noted that the linux kernel itself doesn't seem to care so much about the load average accuracy, and if this data is only used to actually print some stats to the user we could as well tolerate some (rare) inaccuracy.

You definitely have a better big picture of the whole system so I will rely on your expertise :)

Reply:
I merged those methods together. It does make more sense